rust/transaction-filter/src/filter_operator.rs

```diff
@@ -0,0 +1,336 @@
+        match self {
+            APIFilter::TransactionRootFilter(filter) => filter.is_allowed(txn),
+            APIFilter::UserTransactionRequestFilter(ut_filter) => txn
+                .txn_data
+                .as_ref()
+                .map(|txn_data| {
+                    if let TxnData::User(u) = txn_data {
+                        u.request
+                            .as_ref()
+                            .map(|req| ut_filter.is_allowed(req))
+                            .unwrap_or(false)
+                    } else {
+                        false
+                    }
+                })
+                .unwrap_or(false),
+            APIFilter::EventFilter(events_filter) => {
+                if let Some(txn_data) = &txn.txn_data {
+                    let events = match txn_data {
+                        TxnData::BlockMetadata(bm) => &bm.events,
+                        TxnData::Genesis(g) => &g.events,
+                        TxnData::StateCheckpoint(_) => return false,
+                        TxnData::User(u) => &u.events,
+                        TxnData::Validator(_) => return false,
+                    };
+                    events_filter.is_allowed_vec(events)
+                } else {
+                    false
+                }
+            },
+            APIFilter::WriteSetChangeFilter(changes_filter) => {
+                changes_filter.is_allowed_opt_vec(&txn.info.as_ref().map(|inf| &inf.changes))
+            },
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum FilterOperator {
+    And(LogicalAnd),
+    Or(LogicalOr),
+    Filter(APIFilter),
+}
+
+impl FilterOperator {
+    pub fn new_and(and: Vec<FilterOperator>) -> Self {
+        FilterOperator::And(LogicalAnd { and })
+    }
+
+    pub fn new_or(or: Vec<FilterOperator>) -> Self {
+        FilterOperator::Or(LogicalOr { or })
+    }
+
+    pub fn new_filter(filter: APIFilter) -> Self {
+        FilterOperator::Filter(filter)
+    }
+}
+
+impl Filterable<Transaction> for FilterOperator {
+    fn is_valid(&self) -> Result<(), Error> {
+        match self {
+            FilterOperator::And(and) => and.is_valid(),
+            FilterOperator::Or(or) => or.is_valid(),
+            FilterOperator::Filter(filter) => filter.is_valid(),
+        }
+    }
+
+    fn is_allowed(&self, item: &Transaction) -> bool {
+        match self {
+            FilterOperator::And(and) => and.is_allowed(item),
+            FilterOperator::Or(or) => or.is_allowed(item),
+            FilterOperator::Filter(filter) => filter.is_allowed(item),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct LogicalAnd {
+    and: Vec<FilterOperator>,
+}
+
+impl Filterable<Transaction> for LogicalAnd {
+    fn is_valid(&self) -> Result<(), Error> {
+        for filter in &self.and {
+            filter.is_valid()?;
+        }
+        Ok(())
+    }
+
+    fn is_allowed(&self, item: &Transaction) -> bool {
```

Comment:
Good old Rust, feels like these tiny is_allowed functions are the real meat and potatoes and everything else is boilerplate hahah.